rust/rsmgp-sys/src/vertex/mod.rs

```diff
@@ -0,0 +1,194 @@
+    ptr: *mut mgp_vertices_iterator,
+    is_first: bool,
+    context: Memgraph,
+}
+
+impl Default for VerticesIterator {
+    fn default() -> Self {
+        Self {
+            ptr: std::ptr::null_mut(),
+            is_first: true,
+            context: Memgraph {
+                ..Default::default()
+            },
+        }
+    }
+}
+
+impl Drop for VerticesIterator {
+    fn drop(&mut self) {
+        unsafe {
+            if !self.ptr.is_null() {
+                ffi::mgp_vertices_iterator_destroy(self.ptr);
+            }
+        }
+    }
+}
+
+impl Iterator for VerticesIterator {
+    type Item = Vertex;
+
+    fn next(&mut self) -> Option<Vertex> {
+        unsafe {
+            let data: *const mgp_vertex;
+            if self.is_first {
+                self.is_first = false;
+                data = ffi::mgp_vertices_iterator_get(self.ptr);
+            } else {
+                data = ffi::mgp_vertices_iterator_next(self.ptr);
+            }
+
+            if data.is_null() {
+                None
+            } else {
+                let copy_ptr = ffi::mgp_vertex_copy(data, self.context.memory());
+                if copy_ptr.is_null() {
+                    panic!("Unable to allocate new vertex during vertex iteration.");
+                }
+                Some(Vertex {
+                    ptr: copy_ptr,
+                    context: self.context.clone(),
+                })
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Vertex {
+    pub ptr: *mut mgp_vertex,
+    pub context: Memgraph,
+}
+
+impl Drop for Vertex {
+    fn drop(&mut self) {
+        unsafe {
+            if !self.ptr.is_null() {
+                ffi::mgp_vertex_destroy(self.ptr);
+            }
+        }
+    }
+}
+
+impl Vertex {
+    pub fn id(&self) -> i64 {
+        unsafe { ffi::mgp_vertex_get_id(self.ptr).as_int }
+    }
+
+    pub fn labels_count(&self) -> u64 {
+        unsafe { ffi::mgp_vertex_labels_count(self.ptr) }
+    }
+
+    pub fn label_at(&self, index: u64) -> MgpResult<CString> {
+        unsafe {
+            let c_label = ffi::mgp_vertex_label_at(self.ptr, index);
+            if c_label.name.is_null() {
+                return Err(MgpError::OutOfBoundLabelIndexError);
+            }
+            create_cstring(c_label.name)
+        }
+    }
+
+    pub fn has_label(&self, name: &CStr) -> bool {
+        unsafe {
+            let c_mgp_label = mgp_label {
+                name: name.as_ptr(),
+            };
+            ffi::mgp_vertex_has_label(self.ptr, c_mgp_label) != 0
+        }
+    }
+
+    pub fn property(&self, name: &CStr) -> MgpResult<Property> {
+        let unable_alloc_msg = c_str!("Unable to allocate vertex property.");
+        unsafe {
+            let mgp_value =
+                ffi::mgp_vertex_get_property(self.ptr, name.as_ptr(), self.context.memory());
+            if mgp_value.is_null() {
+                ffi::mgp_result_set_error_msg(self.context.result(), unable_alloc_msg.as_ptr());
```

Comment:
I think the Error should be propagated to the function that produced it, and their job should be to catch it and send it to the error_result.